events/api/src/main/java/org/projectnessie/events/api/Event.java

```diff
@@ -0,0 +1,119 @@
+/*
+ * Copyright (C) 2023 Dremio
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.projectnessie.events.api;
+
+import com.fasterxml.jackson.annotation.JsonSubTypes;
+import com.fasterxml.jackson.annotation.JsonTypeInfo;
+import com.google.errorprone.annotations.CanIgnoreReturnValue;
+import java.time.Instant;
+import java.util.Collections;
+import java.util.Map;
+import java.util.Optional;
+import java.util.UUID;
+import org.immutables.value.Value;
+
+/**
+ * Base interface for all events produced by Nessie.
+ *
+ * @see CommitEvent
+ * @see MergeEvent
+ * @see TransplantEvent
+ * @see ReferenceCreatedEvent
+ * @see ReferenceUpdatedEvent
+ * @see ReferenceDeletedEvent
+ * @see ContentStoredEvent
+ * @see ContentRemovedEvent
+ */
+@JsonSubTypes({
+  @JsonSubTypes.Type(name = "COMMIT", value = ImmutableCommitEvent.class),
+  @JsonSubTypes.Type(name = "MERGE", value = ImmutableMergeEvent.class),
+  @JsonSubTypes.Type(name = "TRANSPLANT", value = ImmutableTransplantEvent.class),
+  @JsonSubTypes.Type(name = "REFERENCE_CREATED", value = ImmutableReferenceCreatedEvent.class),
+  @JsonSubTypes.Type(name = "REFERENCE_UPDATED", value = ImmutableReferenceUpdatedEvent.class),
+  @JsonSubTypes.Type(name = "REFERENCE_DELETED", value = ImmutableReferenceDeletedEvent.class),
+  @JsonSubTypes.Type(name = "CONTENT_STORED", value = ImmutableContentStoredEvent.class),
+  @JsonSubTypes.Type(name = "CONTENT_REMOVED", value = ImmutableContentRemovedEvent.class)
+})
+@JsonTypeInfo(use = JsonTypeInfo.Id.NAME, property = "type")
+public interface Event {
+
+  /** The type of the event. */
+  EventType getType();
+
+  /**
+   * The id of the event. The UUID is generated by the event emitter. No assumptions should be made
+   * about the version of the UUID. Most UUIDs will be version 4 (random).
```

Comment:
tangential: Would it make sense to use time UUIDs since we already have `getCreatedAt()`. Would time UUIDs be preferable to reduce the strain of the source of randomness in the server?

Reply:
For that matter, WDYT about simply removing `Most UUIDs will be version 4 (random)` from the javadoc? :)

Reply:
Not sure how well time-UUIDs work in a container environment. They get their "unique" part from the encoded ethernet address, which was supposed to be a hardware thing. But with containers, you don't have that - or cannot assume that some ethernet address is "uniquely assigned by a hardware vendor". So pure-random UUIDs are probably fine. Further, it requires a lot of synchronisation to get time-UUIDs right - not just within a process, but strictly speaking across processes (think: _time_-UUID).

But +1 on removing any assumption on the UUID type.

Reply:
...and for completeness, this is one of the few generators that handle version 6 and 7:

https://github.com/f4b6a3/uuid-creator

Reply:
As for the MAC part, I believe the spec allows generating it randomly (which we'd have to do at most once per start).

Reply:
Certainly a separate issue for time UUIDs :+1:

Reply:
There you go: https://github.com/projectnessie/nessie/issues/6689.